packages/engine-server/src/workspace.ts

```diff
@@ -425,39 +451,49 @@ export class WorkspaceService {
     const out = await Promise.all(
       allRepos.map(async (root) => {
         const git = new Git({ localUrl: root });
-        if (await git.hasRemote()) {
+        // It's impossible to pull if there is no remote, or if there are tracked files that have changes
+        if (
+          !(await git.hasRemote()) ||
+          (await git.hasChanges({ untrackedFiles: "no" }))
+        )
+          return undefined;
+        if (!this.shouldWorkspaceVaultSync("pull", root)) return undefined;
+        try {
           await git.pull();
-          return root;
+        } catch (err) {
+          const stderr = err.stderr ? `: ${err.stderr}` : "";
+          throw new DendronError({
+            message: `error pulling vault${stderr}`,
+            payload: { err, repoPath: root },
+          });
         }
-        return undefined;
+        return root;
       })
     );
     return _.filter(out, (ent) => !_.isUndefined(ent)) as string[];
   }
 
   /** Returns the list of vaults that were attempted to be pushed, even if there was nothing to push. */
   async pushVaults(): Promise<string[]> {
-    const allRepos = await this.getAllRepos();
     const vaults = this.config.vaults;
-    const wsRoot = this.wsRoot;
+    const vaultsPushed = new Set<string>();
     const out = await Promise.all(
-      allRepos.map(async (root) => {
+      vaults.map(async (vault) => {
+        const root = await this.getVaultRepo(vault);
+        if (root === "") return undefined; // The vault is not in a repository
```

Comment:
since our convention has been to return undefined for invalid cases, we should stick to that. could you also add this to the [docs](https://wiki.dendron.so/notes/773e0b5a-510f-4c21-acf4-2d1ab3ed741e.html#summary) ?